src/moonlink_service/src/error.rs

```diff
@@ -1,15 +1,100 @@
-#[derive(Debug, thiserror::Error)]
+use arrow_schema::ArrowError;
+use moonlink_error::{ErrorStatus, ErrorStruct};
+use std::io;
+use std::result;
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error)]
 pub enum Error {
-    #[error("Arrow schema error: {0}")]
-    ArrowSchema(#[from] arrow_schema::ArrowError),
-    #[error("Backend error: {0}")]
-    Backend(#[from] moonlink_backend::Error),
-    #[error("IO error: {0}")]
-    Io(#[from] std::io::Error),
-    #[error("RPC error: {0}")]
-    Rpc(#[from] moonlink_rpc::Error),
-    #[error("Join error: {0}.")]
-    TokioTaskJoin(#[from] tokio::task::JoinError),
-}
-
-pub type Result<T> = std::result::Result<T, Error>;
+    #[error("{0}")]
+    Arrow(ErrorStruct),
+
+    #[error("{0}")]
+    Backend(ErrorStruct),
+
+    #[error("{0}")]
+    Io(ErrorStruct),
+
+    #[error("{0}")]
+    Rpc(ErrorStruct),
+
+    #[error("{0}")]
+    TaskJoin(ErrorStruct),
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+impl From<ArrowError> for Error {
+    #[track_caller]
+    fn from(source: ArrowError) -> Self {
+        let status = match source {
+            ArrowError::IoError(_, _) => ErrorStatus::Temporary,
+
+            // All other errors are regard as permanent
+            _ => ErrorStatus::Permanent,
+        };
+
+        Error::Arrow(ErrorStruct::new(format!("Arrow error: {source}"), status).with_source(source))
```

Comment:
`source` here is duplicate at both string message, also as a separate source field, don't think we need it twice.